src/pinata/mod.rs

```diff
@@ -0,0 +1,59 @@
+use std::{env, path::Path};
+
+use dotenvy::dotenv;
+use pinata_sdk::{PinByFile, PinataApi};
+
+use crate::upload::UploadError;
+
+pub trait PinataClient: Sized {
+    fn new() -> impl std::future::Future<Output = Result<Self, UploadError>> + Send;
+    fn upload_file_to_ipfs(
+        &self,
+        path: &Path,
+    ) -> impl std::future::Future<Output = Result<String, UploadError>> + Send;
+}
+
+pub struct PinataClientImpl {
+    pinata_api: PinataApi,
+}
+
+impl PinataClient for PinataClientImpl {
+    async fn new() -> Result<Self, UploadError> {
+        dotenv().ok();
+        match (env::var("PINATA_API_KEY"), env::var("PINATA_API_SECRET")) {
+            (Ok(api_key), Ok(secret_api_key)) => {
+                let api = PinataApi::new(api_key, secret_api_key)
+                    .map_err(|_| UploadError::Authentication)?;
+                api.test_authentication()
+                    .await
+                    .map_err(|_| UploadError::Authentication)?;
+                Ok(PinataClientImpl { pinata_api: api })
+            }
+            _ => Err(UploadError::Ipfs),
+        }
+    }
+
+    /// Uploads a file at the given path to a Pinata IPFS gateway.
+    async fn upload_file_to_ipfs(&self, path: &Path) -> Result<String, UploadError> {
+        match self
+            .pinata_api
+            .pin_file(PinByFile::new(path.to_string_lossy()))
+            .await
+        {
+            Ok(pinned_object) => Ok(pinned_object.ipfs_hash),
+            Err(_) => Err(UploadError::Ipfs),
+        }
+    }
+}
+
+pub struct MockPinataClient;
```

Comment:
Maybe a comment here explaining the intended purpose of the Mock Client?

Reply:
Also this is only used for testing; although minimal - this code would still get compiled into the release binary.
Hence maybe worthwhile creating and putting this within the test macro/mod or co-located with the tests.

Reply:
done in 5247178ba449153e2c80057dfad266cc0c14ae44